src/v/config/throughput_control_group.cc

```diff
@@ -82,20 +108,36 @@ operator<<(std::ostream& os, const throughput_control_group& tcg) {
       "{{group_name: {}, client_id: {}, throughput_limit_node_in_bps: {}, "
       "throughput_limit_node_out_bps: {}}}",
       tcg.name,
-      tcg.client_id_re ? tcg.client_id_re->pattern() : ""s,
+      tcg.client_id_matcher,
       tcg.throughput_limit_node_in_bps,
       tcg.throughput_limit_node_out_bps);
     return os;
 }
 
 bool throughput_control_group::match_client_id(
-  const ss::sstring& client_id) const {
-    if (!client_id_re) {
+  const std::optional<std::string_view> client_id) const {
+    if (!client_id_matcher) {
         // omitted match criterion means "always match"
         return true;
     }
-    return re2::RE2::FullMatch(
-      re2::StringPiece(client_id.c_str(), client_id.length()), *client_id_re);
+    return std::visit(
+      [client_id](auto&& v) -> bool {
+          using T = std::decay_t<decltype(v)>;
+          if constexpr (std::is_same_v<
+                          T,
+                          client_id_matcher_type::unspecified_tag>) {
+              // only missing client_id matches the unspecified tag
+              return !client_id;
+          } else if constexpr (std::is_same_v<T, copyable_RE2>) {
+              // missing client_id never matches a re
+              return client_id
+                     && re2::RE2::FullMatch(re2::StringPiece(*client_id), v);
+          } else {
+              static_assert(
+                always_false_v<T>, "non-exhaustive client_id_re_type visitor");
+          }
+      },
```

Comment:
this could be `if std::holds_alternative` instead of using visit with constexpr etc... but maybe using optional woudl be even simpler?

Reply:
I've compared the above code with `holds_alternative` approach:
```
    if (std::holds_alternative<client_id_matcher_type::unspecified_tag>(
          client_id_matcher->v)) {
        // only missing client_id matches the unspecified tag
        return !client_id;
    }
    if (auto* const v = std::get_if<copyable_RE2>(&client_id_matcher->v); v) {
        // missing client_id never matches a re
        return client_id
               && re2::RE2::FullMatch(re2::StringPiece(*client_id), *v);
    }
    assert(false, "non-exhaustive client_id_re_type visitor");
```

So it's a comparable amount of code, a bit simpler, but it has an important drawback: a runtime fallback case. With `visit`, a missed case will cause compilation to fail, but with `holds_alternative` it will be a runtime assert/soft assert.

Reply:
> maybe using optional woudl be even simpler?

+1